File: app/src/main/java/me/yxp/qfun/hook/api/OnTroopJoin.java
package me.yxp.qfun.hook.api;

import org.json.JSONObject;

import java.lang.reflect.Method;
import java.util.ArrayList;

import kotlin.collections.CollectionsKt;
import me.yxp.qfun.hook.base.ApiHookItem;
import me.yxp.qfun.utils.dexkit.DexKit;
import me.yxp.qfun.utils.hook.HookUtils;
import me.yxp.qfun.utils.json.ProtoData;
import me.yxp.qfun.utils.qq.HostInfo;
import me.yxp.qfun.utils.qq.QQUtils;
import me.yxp.qfun.utils.reflect.ClassUtils;
import me.yxp.qfun.utils.reflect.MethodUtils;

public final class OnTroopJoin extends ApiHookItem {
    public static final ApiHookItem INSTANCE = new OnTroopJoin();

    @Override
    public void loadHook() throws Throwable {
        if (HostInfo.isTIM()) {
            Method handleMemberJoin = DexKit.getMethod("OnTroopJoin");
            HookUtils.hookAlways(handleMemberJoin, null, param -> handleJoin((String) param.args[0], (String) param.args[1]));
        }

        if (HostInfo.isQQ()) {
            Class<?> troopMemberAddPushProcessor = ClassUtils
                    .load("com.tencent.qqnt.push.processor.TroopMemberAddPushProcessor");
            Method joinGroup = MethodUtils.create(troopMemberAddPushProcessor)
                    .withReturnType(void.class)
                    .withParamTypes(ArrayList.class)
                    .findOne();

            HookUtils.hookAlways(joinGroup, null, param -> {
                ArrayList<Byte> byteList = (ArrayList<Byte>) param.args[0];
                ProtoData data = new ProtoData();
                data.fromBytes(CollectionsKt.toByteArray(byteList));
                JSONObject json = data.toJSON();

                String troopUin = String.valueOf(json.getJSONObject("1").getLong("1"));
                String memberUid = json.getJSONObject("3").getJSONObject("2").getString("3");
                String memberUin = QQUtils.getUinFromUid(memberUid);
                handleJoin(troopUin, memberUin);
            });
        }


    }

    private void handleJoin(String troopUin, String memberUin) throws Throwable {
        for (Listener listener : mListenerSet) {
            if (listener instanceof TroopJoinListener) {
                ((TroopJoinListener) listener).onJoin(troopUin, memberUin);
            }
        }
    }

    public interface TroopJoinListener extends Listener {
        void onJoin(String troopUin, String memberUin) throws Throwable;
    }
}

File: app/src/main/java/me/yxp/qfun/hook/api/OnTroopJoin.kt
package me.yxp.qfun.hook.api

import com.tencent.mobileqq.troop.onlinepush.api.impl.TroopOnlinePushHandler
import com.tencent.qqnt.troopmemberlist.ITroopMemberListRepoApi
import kotlinx.coroutines.suspendCancellableCoroutine
import me.yxp.qfun.annotation.HookItemAnnotation
import me.yxp.qfun.common.ModuleScope
import me.yxp.qfun.hook.base.BaseApiHookItem
import me.yxp.qfun.hook.base.Listener
import me.yxp.qfun.utils.dexkit.DexKitTask
import me.yxp.qfun.utils.hook.getFirstArg
import me.yxp.qfun.utils.hook.hookAfter
import me.yxp.qfun.utils.json.ProtoData
import me.yxp.qfun.utils.json.str
import me.yxp.qfun.utils.json.walk
import me.yxp.qfun.utils.qq.HostInfo
import me.yxp.qfun.utils.qq.api
import me.yxp.qfun.utils.reflect.findMethod
import me.yxp.qfun.utils.reflect.toClass
import org.luckypray.dexkit.query.FindMethod
import org.luckypray.dexkit.query.base.BaseMatcher
import kotlin.coroutines.resume

@HookItemAnnotation("监听用户入群")
object OnTroopJoin : BaseApiHookItem<TroopJoinListener>(), DexKitTask {
    override fun loadHook() {

        if (HostInfo.isTIM) {
            requireMethod("handleJoin").hookAfter(this) { param ->
                handleJoin(
                    param.args[0] as String, param.args[1] as String
                )
            }
        }

        if (HostInfo.isQQ) {
            "com.tencent.qqnt.push.processor.TroopMemberAddPushProcessor".toClass.findMethod {
                returnType = void
                paramTypes(arrayList)
            }.hookAfter(this) { param ->
                val byteList = param.getFirstArg<ArrayList<Byte>>()
                val json = ProtoData().apply {
                    fromBytes(byteList?.toByteArray() ?: byteArrayOf())
                }.toJSON()
                val troopUin = json.walk("3", "2", "1").toString()
                val memberUid = json.walk("3", "2", "3").str ?: return@hookAfter

                ModuleScope.launchIO(name) {
                    var uin = ""
                    repeat(10) {
                        uin = getUinFromUid(memberUid)
                        if (uin.isNotEmpty()) return@repeat
                    }
                    handleJoin(troopUin, uin)
                }

            }
        }


    }

    private fun handleJoin(troopUin: String, memberUin: String) {
        forEachChecked { it.onJoin(troopUin, memberUin) }
    }

    private suspend fun getUinFromUid(uid: String): String =
        suspendCancellableCoroutine { continuation ->
            api<ITroopMemberListRepoApi>().fetchTroopMemberUin(uid) { b, s ->
                continuation.resume(if (b) s else "")
            }
        }


    override fun getQueryMap(): Map<String, BaseMatcher> {
        val clazz = String::class.java
        return mapOf(
            "handleJoin" to FindMethod().apply {
                matcher {
                    declaredClass(TroopOnlinePushHandler::class.java)
                    returnType(Void.TYPE)
                    paramTypes(clazz, clazz, clazz)
                    usingStrings("handleMemberAdd addMemberUin:")
                }
            })
    }
}

fun interface TroopJoinListener : Listener {
    fun onJoin(troopUin: String, memberUin: String)
}